File: session/session.php
<?php


namespace Session;

$rootDir = __DIR__.'/..';

// composer
require_once $rootDir.'/vendor/autoload.php';


require_once $rootDir.'/src/Ething.php';
$ething = new \Ething\Ething();




/*
* session management
*/

function isAuthenticated($checkCsrf = null) {
	global $ething;
	
	$cookieName = $ething->config('session.cookie_name');
	
	if( !empty($_COOKIE[$cookieName]) && !empty($_COOKIE['Csrf-token']) && (!$checkCsrf || checkCsrf()) ){
			
		// ok
		
		// decode the JWT token !
		try {
			$sessionData = \Firebase\JWT\JWT::decode($_COOKIE[$cookieName], $ething->config('session.secret'), array('HS256'));
			if(isset($sessionData->sessionId)){
				return $sessionData;
			}
		} catch( \Exception $e ){ /* invalid JWT */ }
		
		
	}
	
	return false;
}

function checkCsrf() {
	// Csrf control
	$csrfToken = isset($_COOKIE['Csrf-token']) ? $_COOKIE['Csrf-token'] : null;
	if( $csrfToken ){
		
		$csrfToken2 = isset($_GET['csrf_token']) ? $_GET['csrf_token'] : null; // check in the query string
		if(!$csrfToken2)
			$csrfToken2 = isset($_SERVER['HTTP_X_CSRF_TOKEN']) ? $_SERVER['HTTP_X_CSRF_TOKEN'] : null; // check in the header
		
		return $csrfToken === $csrfToken2;
	}
	return false;
}

function authenticate($password) {
	global $ething;
	
	if($ething->config('auth.localonly') && !\Ething\Helpers::isLocalClient()){
		return false;
	}
	
	if( $ething->config('auth.password') === $password ){
	
		// set session cookie (httponly)
		$expireAt = time () + intval($ething->config('session.expiration'));
		$path = preg_replace('#/session/.*$#','',$_SERVER['REQUEST_URI']);
		$secure = isset($_SERVER['HTTPS']) && 'on' === $_SERVER['HTTPS'];
		$sessionData = array(
			'sessionId' => uniqid('',true),
			'iat' => time(),
			'exp' => $expireAt
		);
		$csrf_token = hash_hmac("md5", $sessionData['sessionId'], $ething->config('session.secret'));
		
		$jwt = \Firebase\JWT\JWT::encode($sessionData, $ething->config('session.secret'));
		
		setCookie($ething->config('session.cookie_name'), $jwt, $expireAt, $path, null, $secure, true);
		setCookie('Csrf-token', $csrf_token, $expireAt, $path, null, $secure, false);
		
		return (object)$sessionData;
	} else {
		return false;
	}
}

function unauthenticate() {
	global $ething;
	$path = preg_replace('#/session/.*$#','',$_SERVER['REQUEST_URI']);
	// remove the session cookie
	$cookieSessionName = $ething->config('session.cookie_name');
	setcookie($cookieSessionName, "", time() - 3600, $path);
	setcookie('Csrf-token', "", time() - 3600, $path);
	unset($_COOKIE[$cookieSessionName]);
	unset($_COOKIE['Csrf-token']);
}

File: ething/webserver/session.py
# coding: utf-8

import time
import re
import hmac
import hashlib
import jwt


def uniqid(prefix=''):
    return prefix + hex(int(time.time()))[2:10] + hex(int(time.time()*1000000) % 0x100000)[2:7]


class Session(object):

    def __init__(self, core):
        self.core = core

    def authenticate(self, login, password, request, response):
        # if self.core.config['auth']['localonly']:
        #    pass

        if login == self.core.config['auth']['username'] and password == self.core.config['auth']['password']:

            self.generate(request, response)

            return True

        return False

    def generate(self, request, response, sessionData=None):
        # set session cookie (httponly)
        expireAt = int(time.time() + self.core.config['session']['expiration'])

        path = re.sub('/auth/.*$', '', str(request.url_rule))
        if not path:
            path = '/'

        secure = bool(request.url.startswith('https://'))

        if sessionData is None:
            sessionData = {
                'sessionId': uniqid(''),
                'iat': time.time()
            }

        sessionData['exp'] = expireAt

        csrf_token = hmac.new(self.core.config['session']['secret'].encode(
            'utf8'), sessionData['sessionId'].encode('utf8'), hashlib.md5).hexdigest()

        token = jwt.encode(sessionData, self.core.config['session']['secret'])

        try:
            response.set_cookie(self.core.config['session']['cookie_name'], token,
                                expires=expireAt, domain=request.host, path=path, secure=secure, httponly=True)
            response.set_cookie('Csrf-token', csrf_token, expires=expireAt,
                                domain=request.host, path=path, secure=secure, httponly=False)
        except:
            # when setting a cookie for domain==localhost, an exception is raised : "Setting 'domain' for a cookie on a server running locally (ex: localhost) is not supported by complying browsers. You should have something like: '127.0.0.1 localhost dev.localhost' on your hosts file and then point your server to run on 'dev.localhost' and also set 'domain' for 'dev.localhost'"
            pass

    def refresh(self, request, response):

        token = self.get_token(request)

        if token:

            try:
                sessionData = jwt.decode(
                    token, self.core.config['session']['secret'])
            except jwt.exceptions.InvalidTokenError:
                return False

            self.generate(request, response, sessionData)

            return True

        return False

    def checkCsrf(self, request):
        # Csrf control

        csrfToken = request.cookies.get('Csrf-token')

        if csrfToken:
            # check in the query string
            csrfToken2 = request.args.get('Csrf-token')
            if not csrfToken2:
                csrfToken2 = request.headers.get('HTTP_X_CSRF_TOKEN')

            return csrfToken == csrfToken2

        return False

    def get_token(self, request):
        return request.cookies.get(self.core.config['session']['cookie_name'])

    def isAuthenticated(self, request, checkCsrf=None):

        token = self.get_token(request)

        if token and ((not checkCsrf) or self.checkCsrf(request)):
            # ok decode the JWT token !

            try:
                sessionData = jwt.decode(
                    token, self.core.config['session']['secret'])

                if 'sessionId' in sessionData:
                    return sessionData

            except:
                # invalid JWT
                pass

        return False

    def unauthenticate(self, request, response):

        path = re.sub('/auth/.*$', '', str(request.url_rule))
        if not path:
            path = '/'

        # remove the session cookie
        try:
            response.set_cookie(
                self.core.config['session']['cookie_name'], '', expires=0, domain=request.host, path=path)
            response.set_cookie('Csrf-token', '', expires=0,
                                domain=request.host, path=path)
        except:
            # when setting a cookie for domain==localhost, an exception is raised
            pass
